### DISCUSSION TITLE
🐛 AutoMoq fails to work with TimeProvider in .NET 8

### DISCUSSION THREAD

**Discussion by WestDiscGolf** [AutoFixture/AutoFixture#1473] [Q&A]
### Describe the Bug
I am looking to use the `TimeProvider` construct from .NET 8 with AutoMoq functionality. However when a class is dependent on `TimeProvider` it doesn't process. However if you specify a different abstract class it works as expected.

I **think** it's to do with one of the `IRequestSpecification` not functioning as expected but I can't narrow down the issue. Can someone help?

### Scenario
This reproduces the issue.

```
    public class ReproduceAbstractTests
    {
        [Fact]
        public void Abstract()
        {
            var fixture = new Fixture();
            fixture.Customize(new AutoMoqCustomization());

            var mock = fixture.Freeze<Mock<WasHere>>();

            var sut = fixture.Create<Wrapper<WasHere>>();

            // successful
            Assert.Same(mock.Object, sut.Value);
        }

        [Fact]
        public void TimeProvider()
        {
            var fixture = new Fixture();
            fixture.Customize(new AutoMoqCustomization());

            var mock = fixture.Freeze<Mock<TimeProvider>>();

            var sut = fixture.Create<Wrapper<TimeProvider>>();

            // failure
            Assert.Same(mock.Object, sut.Value);
        }
    }

    public abstract class WasHere { }

    public class Wrapper<T>
    {
        public Wrapper(T value)
        {
            Value = value;
        }

        public T Value { get; }
    }
```

### Expected Behavior
The expected Mock which has been frozen will also be used for `Wrapper<T>` construction.

### Screenshots

<img width="522" alt="image" src="https://github.com/user-attachments/assets/395da8ae-d8f2-44f2-ad9c-270806cea1b0">

### Test Failure

Xunit.Sdk.SameException
Assert.Same() Failure
Expected: Mock<System.TimeProvider:5921>.Object
Actual:   SystemTimeProvider { LocalTimeZone = (UTC+00:00) Dublin, Edinburgh, Lisbon, London, TimestampFrequency = 10000000 }
   at AutoFixture.AutoMoq.UnitTest.ReproduceAbstractTests.TimeProvider() in C:\temp\github\AutoFixture\Src\AutoMoqUnitTest\AutoMoqCustomizationTest.cs:line 214
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)

Thanks!

**Comment by aivascu** [comment]
@WestDiscGolf thank you for raising this.

AutoFixture provides a feature where it tries to construct objects by either invoking the constructor or a factory function.
In this case it sees as a factory the public static property `TimeProvider.System` that generates a `TimeProvider` instance.
You can reproduce the same behavior with your original example if you define your custom types like below.

```cs
public abstract class WasHere
{
    public static WasHere Default { get; } = new WasThere();
}

public sealed class WasThere : WasHere;
```

So for now it seems that this is the expected behavior.
I'll investigate why the fixture has not picked up the frozen mock. This is either a builder priority issue or the specification did not understand that it should return the mocked (non-interface) type.

I think that the original implementation did not expect someone to design types like `TimeProvider`. It's bad practice for an abstract type to know of its derivate types.

**Reply by WestDiscGolf** [reply]
Thanks for taking the time to reply.

The factory structure makes sense however crazy the design of TimeProvider is. I agree the abstract type knowing about an implementation is a bit quirky. I thought I was going a bit crazy tbh but at least there is a why as to what was going on 😀

When I was debugging the code last night I felt like I was going around in circles a bit tbh. I think probably due to lack of familiarity with the code base but it felt like one of the "is satisfied" checks possibly wasn't firing as expected?

I appreciate you taking the time to investigate and would be really interested in the solution 😀

**Reply by Kralizek** [reply]
As a workaround, I use the following snippet, based on FakeItEasy. 

```
fixture.Register(A.Fake<TimeProvider>);
```

The equivalent for Moq would be:

```
fixture.Register(Mock.Of<TimeProvider>());
```

**Reply by aivascu** [reply]
@WestDiscGolf the reason the abstract type is not picked up is because the mock relay is at the end of the pipeline (ResidueCollectors) while the constructor is at the middle of the pipeline. So AutoFixture finds a way to construct the instance of the abstract class before it gets to attempt and resolve a mock for it.

A workaround as @Kralizek mentioned is to inject/register the mock instance repeatedly so it gets added to the front of the pipeline.

I'm honestly not sure how to proceed in this situation, because changing this behavior (i.e. bringing the mock relay to the front) will break existing functionality. AutoFixture is intended not to appeal to brownfield code, and enforce design best practices, yet Microsoft keeps shoving out poorly designed features that are the equivalent of "fan service".

**Reply by Kralizek** [reply]
Given the importance of TimeProvider, I'd consider adding a relay to the default implementation as singleton in the Fixture constructor.

I am not sure if the TP is ASP.NET Core specific, but an idea would be to create a AutoFixture.AspNetCore package that contains all the aspnet core specific customizations for users to pick up.

**Reply by aivascu** [reply]
That can save bit of performance by not executing the reflection query on the type, but I don't think that will have any impact on how the mock is resolved. Consider the following example.

```cs
public abstract class TypeWithVirtualMembers
{
    public virtual void Method()
    {
    }
}

[Fact]
public void TestTypeWithVirtualMembers()
{
    var fixture = new Fixture().Customize(new AutoMoqCustomization());
    var mock = fixture.Freeze<Mock<TypeWithVirtualMembers>>();

    var sut = fixture.Create<TypeWithVirtualMembers>();

    Assert.Same(mock.Object, sut);
}
```
In this scenario there is no problem, the mock will be resolved properly because there is no other way to construct the type.
However if we make `TypeWithVirtualMembers` a concrete type, the test will fail because the pipeline has found a way to create the type before it got to the residue collectors. Basically mocks of concrete types or types that can be resolved by usual means are not resolved via the mock relay.

If we'd add the mock relay to the front then it will attempt to resolve a mock for all types, which will result mock being generated for all types and a significant performance hit.

It would be nice if we could handle the freezing because then we could additionally inject a fixed builder at the front of the pipeline, however that's not possible.

Another option would be to use a parameter customization attribute but that means that the test framework library needs to know of the mock library which is fine for concrete test project but not for a distributable package.

**Reply by aivascu** [reply]
> I am not sure if the TP is ASP.NET Core specific, but an idea would be to create a AutoFixture.AspNetCore package that contains all the aspnet core specific customizations for users to pick up.

TP is a BCL type, so it's not an issue to add it to the core. I like the idea of a `AutoFixture.AspNetCore`. Do you have some examples of types that AutoFixture could support? Probably a discussion for a separate thread.

**Reply by WestDiscGolf** [reply]
Thanks for the update. I appreciate you taking the time to investigate further. The ability to keep backwards compatibility sounds like it could be painful and I can relate to the pain.

The work around works fine. However a lot of the code I work with and write are based off an InlineAutoMoqData type construct (combination of AutoMoqData and InlineData based DataAttribute) and this allows for customization of the Fixture instance but you don't request anything at this point. If I've understood the problem enough there is currently no way to swap relays or add in custom specimen builders to adjust the behavior?

Is there a potential for an options construct to be introduced into the Fixture creation which swaps behavior?

**Reply by aivascu** [reply]
@WestDiscGolf yes basically there's no way to make the fixture pipeline resolve by default frozen mocks of concrete types or any build-able types.
Since you're using auto-mocking  data attribute you can try the following workaround. You can create a parameter customization source attribute that will inject at the front of the pipeline a new mock relay that will look for your specific type and resolve it as a mock.  The attribute below will allow you to freeze a mock and later resolve it as the underlying type.

```cs
[AttributeUsage(AttributeTargets.Parameter)]
public class MockAttribute : Attribute, IParameterCustomizationSource
{
    public ICustomization GetCustomization(ParameterInfo parameter)
    {
        return new MockedParameterCustomization(parameter);
    }

    private class MockedParameterCustomization : ICustomization
    {
        private readonly ParameterInfo parameter;

        public MockedParameterCustomization(ParameterInfo parameter)
        {
            this.parameter = parameter;
        }

        public void Customize(IFixture fixture)
        {
            var parameterType = parameter.ParameterType;
            if (parameterType.IsGenericType && parameterType.GetGenericTypeDefinition().Equals(typeof(Mock<>)))
            {
                var genericArgument = parameterType.GetGenericArguments().Single();
                fixture.Customizations.Add(new MockRelay(new ExactTypeSpecification(genericArgument)));
                return;
            }

            fixture.Customizations.Add(new MockRelay(new ExactTypeSpecification(parameterType)));
        }
    }
}
```

```cs
[Theory]
[InlineAutoMockData(5)]
public void ResolvesMockedInstanceWhenFrozenAsClass(
    [Frozen] int a, [Frozen][Mock] TimeProvider provider,
    Wrapper<TimeProvider> wrapper)
{
    Assert.Same(provider, wrapper.Value);
    Assert.Equal(5, wrapper.A);
}

[Theory]
[InlineAutoMockData(5)]
public void ResolvesMockedInstanceWhenFrozenAsMock(
    [Frozen] int a, [Frozen][Mock] Mock<TimeProvider> provider,
    Wrapper<TimeProvider> wrapper)
{
    Assert.Same(provider.Object, wrapper.Value);
    Assert.Equal(5, wrapper.A);
}
```

**Reply by WestDiscGolf** [reply]
Thanks for the reply. I have tried the above but it did not appear to work for me. Guessing I did something wrong. I matched the `Wrapper` definition to the same as the example and my `InlineAutoMoqDataAttribute` looks like this ...

```
public class InlineAutoMoqDataAttribute : InlineAutoDataAttribute
{
    public InlineAutoMoqDataAttribute(params object[] values)
        : base(new AutoMoqDataAttribute(), values)
    {
    }

    private class AutoMoqDataAttribute : AutoDataAttribute
    {
        public AutoMoqDataAttribute()
            : base(() =>
            {
                var fixture = new Fixture();
                fixture.Customize(new AutoMoqCustomization());
                return fixture;
            })
        {
        }
    }
}
```

I am using version [4.18.1](https://www.nuget.org/packages/AutoFixture.AutoMoq/4.18.1) of all the libraries.

**Reply by aivascu** [reply]
@WestDiscGolf sorry I assumed you're using the preview of v5. The only difference in the provided example is that you'll not be able to inject the inline arguments, the rest should be working the same way.
I created a gist based on the `v4.18.1` for you. https://gist.github.com/aivascu/b946a2fb6021d9648ae9d341ceb1c28b
Let me know if this helped, if not then provide some information on what exactly is not working so I can troubleshoot, preferably a minimum reproducible example.
